src/init.cpp

```diff
@@ -862,9 +866,15 @@ bool AppInit2(boost::thread_group& threadGroup)
     boost::filesystem::path pathLockFile = GetDataDir() / ".lock";
     FILE* file = fopen(pathLockFile.string().c_str(), "a"); // empty lock file; created if it doesn't exist.
     if (file) fclose(file);
-    static boost::interprocess::file_lock lock(pathLockFile.string().c_str());
-    if (!lock.try_lock())
-        return InitError(strprintf(_("Cannot obtain a lock on data directory %s. Bitcoin Core is probably already running."), strDataDir));
+
+    try {
+        static boost::interprocess::file_lock lock(pathLockFile.string().c_str());
+        if (!lock.try_lock())
+            return InitError(strprintf(_("Cannot obtain a lock on data directory %s. Bitcoin Core is probably already running."), strDataDir));
+    } catch(const boost::interprocess::interprocess_exception& e) {
```

Comment:
a) chmod 000 <datadir>

b) if the boost docs are to be believed, this is the only exception it
throws.  i could not find a base exception beyond this.

c) i was wary of adding a translation, but this is a pretty ugly error that
should be explained clearly.  we can't log it (beyond -printtoconsole)
because we're blowing up due to an invalid datadir.  what do you think?